test/integration/apiserver/tracing/tracing_test.go

```diff
@@ -105,93 +84,469 @@ kind: TracingConfiguration
+	lock         sync.Mutex
+	traceFound   chan struct{}
+	expectations traceExpectation
+}
+
+func (t *traceServer) Export(ctx context.Context, req *traceservice.ExportTraceServiceRequest) (*traceservice.ExportTraceServiceResponse, error) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	t.expectations.update(req)
+	// if all expectations are met, notify the test scenario by closing traceFound
+	if t.expectations.met() {
+		select {
+		case <-t.traceFound:
+			// traceFound is already closed
+		default:
+			close(t.traceFound)
+		}
 	}
+	return &traceservice.ExportTraceServiceResponse{}, nil
 }
 
-func containsNodeListSpan(req *traceservice.ExportTraceServiceRequest) bool {
+// resetExpectations is used by a new test scenario to set new expectations for
+// the test server.
+func (t *traceServer) resetExpectations(newExpectations traceExpectation) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	t.traceFound = make(chan struct{})
+	t.expectations = newExpectations
+}
+
+// traceExpectation is an expectation for an entire trace
+type traceExpectation []*spanExpectation
+
+// met returns true if all span expectations the server is looking for have
+// been satisfied.
+func (t traceExpectation) met() bool {
+	if len(t) == 0 {
+		return true
+	}
+	// we want to find any trace ID which all span IDs contain.
+	// try each trace ID met by the first span.
+	possibleTraceIDs := t[0].metTraceIDs
+	for _, tid := range possibleTraceIDs {
+		if t.contains(tid) {
+			return true
+		}
+	}
+	return false
+}
+
+// contains returns true if the all spans in the trace expectation contain the
+// trace ID
+func (t traceExpectation) contains(checkTID string) bool {
+	for _, expectation := range t {
+		if !expectation.contains(checkTID) {
+			return false
+		}
+	}
+	return true
+}
+
+// update finds all expectations that are met by a span in the
+// incoming request.
+func (t traceExpectation) update(req *traceservice.ExportTraceServiceRequest) {
 	for _, resourceSpans := range req.GetResourceSpans() {
 		for _, instrumentationSpans := range resourceSpans.GetScopeSpans() {
 			for _, span := range instrumentationSpans.GetSpans() {
-				if span.Name != "HTTP GET" {
-					continue
-				}
-				for _, attr := range span.GetAttributes() {
-					if attr.GetKey() == "http.url" {
-						value := attr.GetValue().GetStringValue()
-						if strings.HasSuffix(value, "/api/v1/nodes") {
-							// We found our request!
-							return true
-						}
-					}
-				}
+				t.updateForSpan(span)
 			}
 		}
 	}
+}
+
+// updateForSpan updates expectations based on a single incoming span.
+func (t traceExpectation) updateForSpan(span *tracev1.Span) {
+	for i, spanExpectation := range t {
+		if span.Name != spanExpectation.name {
+			continue
+		}
+		if !spanExpectation.attributes.matches(span.GetAttributes()) {
+			continue
+		}
+		if !spanExpectation.events.matches(span.GetEvents()) {
+			continue
+		}
+		t[i].metTraceIDs = append(spanExpectation.metTraceIDs, hex.EncodeToString(span.TraceId[:]))
```

Comment:
i'd return true here, so that in line 468 you can just return after you update.

Reply:
you can have more than one match